Avoid loading Python classes during context initialization

Akirathan opened the pull request:
Fixes #13760

### Pull Request Description

Previously, 12 thousand classes from `com.oracle.graal.python` were loaded during context creation even if no python code was executed. With this PR, only 3 classes are loaded.

### Important Notes

- Avoid unnecessary loading of thousands of classes from `com.oracle.graal.python` by not setting `python` option in https://github.com/enso-org/enso/blob/28bcb2cb25f01bbedf747bd5b19203718e5bed8d/engine/common/src/main/java/org/enso/common/ContextFactory.java#L203-L205
- An empty [PythonResourceProvider](https://github.com/enso-org/enso/blob/61713edc88f23af29e917d64f04c456cfdb5b163/engine/python-resource-provider/src/main/java/org/enso/python/resource/provider/PythonResourceProvider.java) needs to be provided for this machinery to work. See its docs.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] Check startup benchmarks for improvement.
- [X] The documentation has been updated, if necessary.
- [X] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [X] Unit tests have been written where possible.

Commit:
fmt

Commit:
Add empty PythonResourceProvider.

Commit:
python resources points to the parent dir

Akirathan commented:
## How to ensure that python-home resource is loaded at runtime?

Either set `polyglotimpl.TraceInternalResources` system prop to true. See https://github.com/oracle/graal/blob/master/truffle/src/com.oracle.truffle.polyglot/src/com/oracle/truffle/polyglot/InternalResourceRoots.java#L348

or:
```diff
diff --git a/engine/common/src/main/java/org/enso/common/ContextFactory.java b/engine/common/src/main/java/org/enso/common/ContextFactory.java
index 0a0418ec30..5abfc6ed5f 100644
--- a/engine/common/src/main/java/org/enso/common/ContextFactory.java
+++ b/engine/common/src/main/java/org/enso/common/ContextFactory.java
@@ -208,6 +208,9 @@ public final class ContextFactory {
             .allowExperimentalOptions(true)
             .allowAllAccess(true)
             .allowHostAccess(allWithTypeMapping())
+            .option("log.python.com.oracle.graal.python.builtins.Python3Core.level", "FINEST")
+            .option("log.python.com.oracle.graal.python.PythonLanguage.level", "FINEST")
+            .option("log.python.com.oracle.graal.python.runtime.PythonContext.level", "FINEST")
             .option(RuntimeOptions.STRICT_ERRORS, Boolean.toString(strictErrors))
             .option(RuntimeOptions.DISABLE_LINTING, Boolean.toString(disableLinting))
             .option(RuntimeOptions.WAIT_FOR_PENDING_SERIALIZATION_JOBS, "true")
diff --git a/engine/runner/src/main/resources/application.conf b/engine/runner/src/main/resources/application.conf
index eadd3e22e4..c7deca153d 100644
--- a/engine/runner/src/main/resources/application.conf
+++ b/engine/runner/src/main/resources/application.conf
@@ -8,6 +8,7 @@ logging-service {
     akka.stream = error
     org.apache.poi.openxml4j.opc = warn
     Standard.Base.Logging.Progress = error
+    python.com.oracle.graal.python = trace
   }
   appenders = [
       {
```

and make sure there are no WARN messages from `com.oracle.graal.python.PythonLanguage`

Akirathan commented:
## How to ensure no unnecessary Python classes are loaded?

Either simply wait for benchmark results mentioned in #13760, or use JFR:
- Use JFR profile with event name `jdk.ClassLoadingStatistics` enabled:
```
    <event name="jdk.ClassLoadingStatistics">
      <setting name="enabled">true</setting>
      <setting name="period">1000 ms</setting>
    </event>
```
- Run Enso with `JAVA_TOOL_OPTIONS="-XX:+FlightRecorder -XX:StartFlightRecording=filename=flight.jfr,settings=my-profile.jfc"`
- Inspect loaded classes either via JDK Mission Control. or via something like this:
```java
import java.io.IOException;
import java.util.HashSet;
import java.util.Set;
import jdk.jfr.consumer.RecordedClass;
import jdk.jfr.consumer.RecordedEvent;
import jdk.jfr.consumer.RecordedFrame;
import jdk.jfr.consumer.RecordingFile;
import java.nio.file.Path;

class ReadJFR {

  public static void main(String[] args) throws IOException {
    Path jfrPath;
    if (args.length == 0) {
      jfrPath = Path.of("/Users/pavel/dev/enso/flight.jfr");
    } else {
      jfrPath = Path.of(args[0]);
    }
    var pythonClassesLoaded = new HashSet<String>();
    try (RecordingFile recordingFile = new RecordingFile(jfrPath)) {
      while (recordingFile.hasMoreEvents()) {
        var event = recordingFile.readEvent();
        if (event.getEventType().getName().equals("jdk.ClassLoad")) {
          var recordedClass = (RecordedClass) event.getValue("loadedClass");
          var className = recordedClass.getName();
          if (className.startsWith("com.oracle.graal.python")) {
            pythonClassesLoaded.add(className);
          }
        }
      }
    }
    System.out.println("Loaded Python classes (" + pythonClassesLoaded.size() + "): ");
  }
}
```

On develop, the number of loaded classes with prefix `com.oracle.graal.python` is more than 12 thousand, even without executing any Python code. On this PR, only 3 such classes are loaded.

Commit:
Remove unused nested class

Akirathan commented:
Startup benchmarks scheduled at https://github.com/enso-org/enso/actions/runs/17076575633

<blockquote><img src="https://opengraph.githubassets.com/390143d460ef88f1948010df28b95ef742833434324239329e27f7136fe80c1b/enso-org/enso" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Benchmark Standard Libraries · enso-org/enso@37039ec</a></strong></div><div>Enso Analytics is a self-service data prep and analysis platform designed for data teams. - Benchmark Standard Libraries · enso-org/enso@37039ec</div></blockquote>

Akirathan commented:
The startup slowdown is fixed:
<img width="2028" height="599" alt="image" src="https://github.com/user-attachments/assets/6b22c456-4d7c-46d7-990d-84f9e6f03a0e" />

<img width="2025" height="600" alt="image" src="https://github.com/user-attachments/assets/223ef848-0f46-4774-8b52-ce9c310afa98" />

<img width="2026" height="598" alt="image" src="https://github.com/user-attachments/assets/ed33e406-c108-423f-97d9-050be66858a6" />

Commit:
fmt

JaroslavTulach commented on engine/common/src/main/java/org/enso/common/ContextFactory.java:
```diff
@@ -193,8 +200,8 @@ public Context build() {
       }
     }
-    if (pythonHome != null) {
-      options.put("python.PythonHome", pythonHome);
```
Replacing one property by another (as @eregon suggested, thanks) seems like an easy fix.

JaroslavTulach commented:
> ## How to ensure that python-home resource is loaded at runtime?

- Why don't you turn this into a JUnit test?
- Use `ObservedMessage` to collect logs from `log.python.com.oracle.graal.python` & co.
- initialize context
- check there are no _"extraction messages"_
- check there is no _sign of python classloading_
- FYI: tutorial [Advanced Usage of Logging](https://wiki.apidesign.org/wiki/FlowControllingTest#Advanced_usage_of_Logging) in tests

JaroslavTulach reviewed (approved):
- Good to see the startup time regression fixed. 
- I'd add a unit test.

Commit:
Move some functions to test/Base_Tests/Utils

Commit:
enso_run does not support env.

It doesn't work anyway.

Commit:
Add Python_Resource_Spec

Akirathan commented:
> I'd add a unit test

It was a bit tough. But I think I have finally managed to do it in 3f68bc1d692fe9473824df839b906bef1f0c45b6.

Commit:
Merge branch 'develop' into wip/akirathan/13760-no-python-ctx-init

# Conflicts:
#	build.sbt

JaroslavTulach commented:
> > I'd add a unit test
> 
> It was a bit tough. But I think I have finally managed to do it in [3f68bc1](https://github.com/enso-org/enso/commit/3f68bc1d692fe9473824df839b906bef1f0c45b6).

- that's enormously huge piece of Enso code!
- I don't like it is spawning a new process
   - it is slow and running `test/Base_Tests` already takes ages
   - we have `org.graalvm.polyglot` API and we can do better
- I was asking for:

> - Why don't you turn this into a **JUnit** test?
> - Use `ObservedMessage` to collect logs from log.python.com.oracle.graal.python & co.
> - initialize context
> - check there are no "extraction messages"
> - check there is no sign of python classloading

- the checking/processing of the logs is correct in 3f68bc1d692fe9473824df839b906bef1f0c45b6
- however can it be a **JUnit** test?
- can it run in the same JVM?
- can the logs be collected in memory (for example by `ObservedMessage`)

hubertp committed:
Fix errors printed when executing PythonExtract

```
[info] running (fork) org.enso.pyextract.PythonExtract lib/java/python-extract/target/python-resources
[error] WARNING: Unknown module: org.graalvm.truffle specified to --enable-native-access
[error] WARNING: A restricted method in java.lang.System has been called
[error] WARNING: java.lang.System::load has been called by com.oracle.truffle.polyglot.JDKSupport in an unnamed module (file:<HOME>.cache/coursier/v1/https/repo1.maven.org/maven2/org/graalvm/truffle/truffle-api/24.2.0/truffle-api-24.2.0.jar)
[error] WARNING: Use --enable-native-access=ALL-UNNAMED to avoid a warning for callers in this module
[error] WARNING: Restricted methods will be blocked in a future release unless native access is enabled
[info] Compiling lib/java/jna-wrapper/src/main/java/module-info.java with javac
```

was a bit too much for me.

Akirathan commented:
> however can it be a JUnit test? can it run in the same JVM?

It cannot. One important thing that I wanted to test is that the Python's language home is resolved inside `built-distribution`. When JUnit tests run, there is no `built-distribution` yet. For example, `runtime-integration-tests` run with `python-resources.jar` on the module-path, which means that the Python's language home in that case will be resolved to the directory that GraalPy unpacks at runtime. This is not the behavior we desire in our product.

I can see the controversy about putting this [Python_Resource_Spec](https://github.com/enso-org/enso/blob/3f68bc1d692fe9473824df839b906bef1f0c45b6/test/Base_Tests/src/System/Python_Resource_Spec.enso) into Base_Tests. I was thinking that maybe Examples_Tests would be a better place. **But `Examples_Tests` are not executed at all on the CI**. They were disabled in https://github.com/enso-org/enso/pull/12948/commits/2a8c686696782e431b0f0aee9a8faed8da90328c.

TL;DR:
- It is not possible to test this behavior in JUnit test
- Moving it to Examples_Tests is equivalent to not having any tests at all, since Examples_Tests are not executed at all.

Commit:
Include java.lang.Process in test NI

Commit:
lint: remove unused imports

Commit:
Merge remote-tracking branch 'origin/wip/akirathan/13760-no-python-ctx-init' into wip/akirathan/13760-no-python-ctx-init

Commit:
lint: remove unused imports

Commit:
Print entire output in test clue.

This will help to debug Windows CI failure at https://github.com/enso-org/enso/actions/runs/17125894196/job/48586384414?pr=13812#step:14:1229

Akirathan commented:
Increase verbosity of test failure in 3c848cb74a4ddeaa6f7404ce2a0d40905d647a23 to help debug Windows failure at https://github.com/enso-org/enso/actions/runs/17125894196/job/48586384414?pr=13812#step:14:1229.

<blockquote><img src="https://opengraph.githubassets.com/4dce33648719f81f55bae9aa275b388c7e5e3a7bb73015a8ecb4e71eafff9364/enso-org/enso" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Avoid loading Python classes during context initialization · enso-org/enso@a56f618</a></strong></div><div>Enso Analytics is a self-service data prep and analysis platform designed for data teams. - Avoid loading Python classes during context initialization · enso-org/enso@a56f618</div></blockquote>

JaroslavTulach commented:
> **But `Examples_Tests` are not executed at all on the CI**. They were disabled in [2a8c686](https://github.com/enso-org/enso/commit/2a8c686696782e431b0f0aee9a8faed8da90328c).

The answer to this isn't to give up, but ask @4e6 to enable the tests again!

Commit:
simple change to force job restart

4e6 commented:
Examples_Tests are still failing https://github.com/enso-org/enso/pull/13845

Commit:
Fix log split for Windows path

mergify[bot] committed:
Merge branch 'develop' into wip/akirathan/13760-no-python-ctx-init
